File: tablix/tablix_hr/page/notification/notification.py
# Copyright (c) 2019, Tablix and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe.model.document import Document
from frappe import msgprint, _, throw

@frappe.whitelist()
def get_data(name):
	if(name):
		try:
			data = frappe.db.sql("""select subject, date, description from `tabNotification` where name = %s  """,(name), as_dict=True)

			if(data):
				return data

		except Exception as e:
			frappe.throw(_("Error, Please Contact to your System Admin. {0}").format(e))

@frappe.whitelist()
def submit_data(user, ref_no, sign):
	if(user and ref_no):
		try:
			doc = frappe.get_doc("Notification", {"name":ref_no})
			emp = frappe.get_doc("Employee", {"user_id": user})

			if(doc and emp):
				emp.append("document_viewed", {
					"document_name": doc.subject,
					"ref_no": doc.name,
					"document_type": "Notification",
					"date": frappe.utils.nowdate()
				})
				
				if(sign):
					emp.signature = sign
				
				emp.save(ignore_permissions=True)

			msg = ("You confirm to have Read, & Understood the " + doc.subject + " of the company. Attached the doccument for your reference. ")

			my_attachments = [frappe.attach_print(doc.doctype, doc.name, file_name=doc.name, print_format="Notification")]

			frappe.sendmail(recipients=[user], subject=doc.subject, delayed=False, message=msg, attachments=my_attachments)

		except Exception as e:
			print(frappe.get_traceback())
			frappe.throw(_("Error, Please Contact to your System Admin. {0}").format(e))


@frappe.whitelist()
def check_accepted(user, ref_no):
	accept = []
	if(user and ref_no):
		doc = frappe.db.sql("""select accepted from `tabPolicy Employee` where email = %s and parent = %s """,(user, ref_no), as_dict= 1)	
		if(doc):
			if(doc[0]['accepted'] == 1):
				return {"accept": doc[0]['accepted']}
			else:
				frappe.db.sql("""update `tabPolicy Employee` set accepted = 1 where email = %s and parent = %s """,(user, ref_no))
				return {"accept": 0}

File: tablix/tablix_hr/page/notification/notification.js
frappe.pages['notification'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Notification',
		single_column: true
	});

	wrapper.view_document = new erpnext.Notification(wrapper);
}

erpnext.Notification = Class.extend({
	init: function(wrapper) {
		var me = this;
		this.parent = wrapper;
		this.page = this.parent.page;
		setTimeout(function() {
			me.setup(wrapper);
			me.make(wrapper);
		}, 0);
	},

	setup: function(wrapper) {
		var me = this;
		this.elements = {
			layout: $(wrapper).find(".layout-main"),

			approve_btn: wrapper.page.set_primary_action(__("<b>Confirm</b>"),
			function() {
				me.accept_confirm();
			}, "fa fa-refresh"),

			refresh_btn: wrapper.page.set_secondary_action(__("<b>Refersh</b>"),
			function() { me.view_document(frappe.route_options.name); }, "fa fa-refresh"),
		};
	},

	accept_confirm: function(){
		var me = this;
		
		var dialog = new frappe.ui.Dialog({
			title: __('Are you sure you want to confirm this?'),
			fields:[
				{fieldtype:'Section Break', fieldname: 'section', label: __('<p style="text-align:center;"><b>Reviewed, Understood, Recorded</b></p>')},
				{fieldtype:'Check', fieldname:'read', reqd:1, label:'I reviewed the Notification.'},
				{fieldtype:'Check', fieldname:'understand', reqd:1, label:'I understand.'},
			],
		});

		dialog.set_primary_action(__('Confirm'), function() {
			var user_data = dialog.get_values();
			var signature = ""
			if(!user_data.read || !user_data.understand){
				frappe.throw("Please read the Notification carefully & accept to confirm.", title="Review Carefully.");
			} else{
				me.submit_accept(signature);
			}
			dialog.hide();
		});
		dialog.show();
	},

	submit_accept: function(signature){
		var me = this;
		var user = frappe.session.user_email;
		var ref_no = frappe.route_options.name;
		
		if( user && ref_no){
			frappe.call({
				"method": "tablix.tablix_hr.page.notification.notification.check_accepted",
				"args": { "user": user, "ref_no": ref_no },
				"callback": function(r){
					if(r){	
						if(r.message.accept == 0){
							frappe.msgprint("Thanks for your review.", title="Reviewed");
							frappe.call({
								"method": "tablix.tablix_hr.page.notification.notification.submit_data",
								"args": { "user": user, "ref_no": ref_no, "sign": signature }
							});
						} else{
							frappe.msgprint("You Already view this document Notification.", title="Already Reviewed");
						}
					}
				}
			});
		} else{
			frappe.throw("System is facing some problems. Kindly contact to your System Admin.");
		}
		
	},
	
	make: function(wrapper){
		var me = this;
		this.body = $('<div></div>').appendTo(this.page.main);
		var data = "";
		var $container = $(frappe.render_template('notification', data)).appendTo(this.body);

		var doc_name = frappe.route_options.name || "";
		if(doc_name){
			me.view_document(doc_name);
		} else{
			frappe.msgprint("Please follow the correct procedure. <b>GO TO Your Assign Notification and then Click on View Notification.</b>")
		}
	},

	view_document: function(documentname){
		console.log(documentname);
		frappe.call({
			"method": "tablix.tablix_hr.page.notification.notification.get_data",
			"args": { "name": documentname },
			"callback": function(r){
				if(r && r.message){

					var data = r.message[0];
					console.log(data);	
					/* document data */
					if(data){
						$("#subject").html(__(data.subject));
						$("#ref_no").html(__(documentname));
						$("#release_date").html(__(data.date));
		
						const html = `
							<table class="table table-bordered" style="background-color: #f9f9f9;">
								<tbody>
									<tr><td>${data.description}</td></tr>
								</tbody>
							</table>`
						$("#get_des").html(html);
					}
				}
			}
		});
	},
});
